### PR TITLE
feat(ibc): implementing ibc context

### PR DESCRIPTION
**Pull request by ImJeremyHe**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #

**Which docs this PR relation**:

Ref #

**Which toolchain this PR adaption**:

No Breaking Change

**Special notes for your reviewer**:

### CI Switch

ci-runs-only: [Chaos CI, Cargo Clippy, Coverage Test, E2E Tests, Code Format, Unit Tests, Web3 Compatible Tests]

### CODE REVIEW

File: core/ibc/src/adapter.rs [hunk 1]
```diff
@@ -1 +1,57 @@
...
+    + ClientTypeStore
+    + ConnectionEndStore
+    + ConnectionIdsStore
+    + ConsensusStateStore
+    + PacketCommitmentStore
+    + PacketReceiptsSet
+    + PacketAcsStore
+    + RecvSequenceStore
+    + SendSequenceStore
+    + SeqAcksPathStore
```
Comment (KaoImin) [inline core/ibc/src/adapter.rs]: I do not think it is a good idea that abstract each storage structure into a separate trait. IMO the storage schema behavior can be described as `get()`, `set()`, `remove()`, `iter() (optional)`. Different schemas just have different types and codec implementations.

File: core/ibc/src/codec.rs [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+        packet::Sequence,
+    },
+    ics24_host::identifier::ConnectionId,
+};
+use ibc_proto::google::protobuf::Any;
+use ibc_proto::ibc::core::channel::v1::Channel as RawChannelEnd;
+use ibc_proto::ibc::core::connection::v1::ConnectionEnd as RawConnectionEnd;
+use tendermint_proto::Protobuf;
+
+pub trait Codec: Sized {
```
Comment (KaoImin) [inline core/ibc/src/codec.rs]: Rename this trait as `CrossChainCodec`

File: core/ibc/src/codec.rs [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    fn encode(&self) -> Option<Vec<u8>>;
+    fn decode(raw: &[u8]) -> Option<Self>;
```
Comment (KaoImin) [inline core/ibc/src/codec.rs]: The encode and decode function should return `ProtocolResult`.

File: core/ibc/src/codec.rs [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+json_codec!(());
+json_codec!(ClientType);
+json_codec!(Sequence);
+json_codec!(Vec<ConnectionId>);
+protobuf_codec!(AnyClientState, Any);
+protobuf_codec!(AnyConsensusState, Any);
+protobuf_codec!(ChannelEnd, RawChannelEnd);
+protobuf_codec!(ConnectionEnd, RawConnectionEnd);
+bin_codec!(PacketCommitment);
+bin_codec!(AcknowledgementCommitment);
```
Comment (KaoImin) [inline core/ibc/src/codec.rs]: The cross chain codec defination and implementation should move to `protocol/src/codec/crosschain.rs`

File: protocol/src/codec/crosschain.rs [hunk 5]
```diff
@@ -82,6 +107,72 @@ impl Decodable for HashWithDirection {
     }
 }
 
+#[derive(Clone)]
+pub struct IbcWrapper<T: Clone>(pub T);
```
Comment (KaoImin) [inline protocol/src/codec/crosschain.rs]:
```suggestion
pub struct IbcWrapper<T: Message>(pub T);
```

Comment (ImJeremyHe) [reply protocol/src/codec/crosschain.rs]: Not all the `ibc item` implements the prost::Message

File: core/ibc/src/adapter.rs [hunk 6]
```diff
@@ -1 +1,5 @@
+use protocol::traits::IbcCrossChainStorage;
 
+pub trait IbcAdapter: IbcCrossChainStorage {
```
Comment (KaoImin) [inline core/ibc/src/adapter.rs]: The trait should be defined in `protocol/traits/ibc.rs`

**Review by KaoImin — DISMISSED** [review]

**Review by KaoImin — APPROVED** [review]

**Pull request merged**